subsys/config/src/config.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+#include <errno.h>
+
+#include "mbedtls/base64.h"
+
+#include "config/config.h"
+#include "config_priv.h"
+#include <zephyr/types.h>
+
+/* mbedtls-base64 lib encodes data to null-terminated string */
+#define BASE64_ENCODE_SIZE(in_size) ((((((in_size) - 1) / 3) * 4) + 4) + 1)
+
+sys_slist_t conf_handlers;
+
+static u8_t conf_cmd_inited;
+
+void conf_init(void)
+{
+	sys_slist_init(&conf_handlers);
+	conf_store_init();
+
+	if (conf_cmd_inited) {
+		return;
+	}
+
+	conf_cmd_inited = 1;
+}
+
+int conf_register(struct conf_handler *handler)
+{
+	sys_slist_prepend(&conf_handlers, &handler->ch_list);
+	return 0;
+}
+
+/*
+ * Find conf_handler based on name.
+ */
+struct conf_handler *conf_handler_lookup(char *name)
+{
+	struct conf_handler *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conf_handlers, ch, ch_list) {
+		if (!strcmp(name, ch->ch_name)) {
+			return ch;
+		}
+	}
+	return NULL;
+}
+
+/*
+ * Separate string into argv array.
+ */
+int conf_parse_name(char *name, int *name_argc, char *name_argv[])
+{
+	char *tok;
+	char *tok_ptr;
+	char *sep = CONF_NAME_SEPARATOR;
+	int i;
+
+	tok = strtok_r(name, sep, &tok_ptr);
+
+	i = 0;
+	while (tok) {
+		name_argv[i++] = tok;
+		tok = strtok_r(NULL, sep, &tok_ptr);
+	}
+	*name_argc = i;
+
+	return 0;
+}
+
+static struct conf_handler *conf_parse_and_lookup(char *name, int *name_argc,
+						  char *name_argv[])
+{
+	int rc;
+
+	rc = conf_parse_name(name, name_argc, name_argv);
+	if (rc) {
+		return NULL;
+	}
+	return conf_handler_lookup(name_argv[0]);
+}
+
+int conf_value_from_str(char *val_str, enum conf_type type, void *vp,
+			int maxlen)
+{
+	s32_t val;
+	s64_t val64;
+	char *eptr;
+
+	if (!val_str) {
+		goto err;
+	}
+	switch (type) {
+	case CONF_INT8:
+	case CONF_INT16:
+	case CONF_INT32:
+	case CONF_BOOL:
+		val = strtol(val_str, &eptr, 0);
+		if (*eptr != '\0') {
+			goto err;
+		}
+		if (type == CONF_BOOL) {
+			if (val < 0 || val > 1) {
+				goto err;
+			}
+			*(bool *)vp = val;
+		} else if (type == CONF_INT8) {
+			if (val < INT8_MIN || val > UINT8_MAX) {
+				goto err;
+			}
+			*(int8_t *)vp = val;
+		} else if (type == CONF_INT16) {
+			if (val < INT16_MIN || val > UINT16_MAX) {
+				goto err;
+			}
+			*(int16_t *)vp = val;
+		} else if (type == CONF_INT32) {
+			*(s32_t *)vp = val;
+		}
+		break;
+	case CONF_INT64:
+		val64 = strtoll(val_str, &eptr, 0);
+		if (*eptr != '\0') {
+			goto err;
+		}
+		*(s64_t *)vp = val64;
+		break;
+	case CONF_STRING:
+		val = strlen(val_str);
+		if (val + 1 > maxlen) {
+			goto err;
+		}
+		strcpy(vp, val_str);
+		break;
+	default:
+		goto err;
+	}
+	return 0;
+err:
+	return -EINVAL;
+}
+
+int conf_bytes_from_str(char *val_str, void *vp, int *len)
+{
+	int err;
+	int rc;
+
+	err = mbedtls_base64_decode(vp, *len, &rc, val_str, strlen(val_str));
```

Comment:
got it, this is the pain point

can we define a generic function and move the implementation details to that function at least.

Reply:
base64 is just a binary-to-text encoding algorithm. I would hesitate to create such a module - It will be really easy  replace call to mbedtls-base64 to any else bin-to-tex encoding call (once will be available).
As mcumgr (the Go tool) has implemented functionality for r/w configuration it is better now to keep base64 encoding.

Reply:
ok, for now, we will live with this.